### DISCUSSION TITLE
Dataclasses get converted into dicts instead of being validated by pydantic directly

### DISCUSSION THREAD

**Discussion by ImogenBits** [fastapi/fastapi#6079] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from dataclasses import dataclass, field
from typing import Any
from pydantic import BaseModel, BaseConfig
from fastapi import FastAPI


class ThingBase:
    id: int
    circular_ref: "ThingBase"


class ThingID:
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, obj: Any) -> int:
        if isinstance(obj, int):
            return obj
        elif isinstance(obj, ThingBase):
            return obj.id
        else:
            raise TypeError


class Model(BaseModel):
    id: int
    circular_ref: ThingID

    class Config(BaseConfig):
        orm_mode = True


class ThingNormal(ThingBase):
    def __init__(self) -> None:
        super().__init__()
        self.id = 0
        self.circular_ref = self


@dataclass
class ThingDataclass(ThingBase):
    id: int = field(init=False)
    circular_ref: ThingBase = field(init=False)

    def __post_init__(self):
        self.id = 0
        self.circular_ref = self


app = FastAPI()


@app.get("/normal", response_model=Model)
def normal_route():
    return ThingNormal()


@app.get("/dataclass", response_model=Model)
def dataclass_route():
    return ThingDataclass()
```

### Description

My project uses a lot of database models with circular references. When serving them to the web api I'd like them to just get converted to their IDs and I've been using a custom pydantic validator for that (not sure if that's actually what you're supposed to use tbh). The problem comes in now that I wanted to switch the database models to SQLAlchemy's new dataclass ones. When using these (or any other dataclass) FastAPI tries to first convert them into a dict using the default `dataclass.asdict()` function ([here](https://github.com/tiangolo/fastapi/blob/4e298356097a3e1bbe2bf28424e7d5392a9fd96b/fastapi/routing.py#L106)), which can't handle circular references at all and crashes because of a stack overflow.

I'm not really sure if this even is a bug or intended behaviour. Is there a better way to handle this kind of circluar references? Or even more generally, automatically converting the types of some entries of the returned object to the type indicated in the `response_model`?  Either way I think that it'd be more intuitive if `fastapi.routing.serialize_response` would take into account that the field might have `orm_mode` enabled and wouldn't then preprocess the object into a dict.  Currently this only happens if the passed object has `orm_mode` enabled, not the field.

### Operating System

Windows

### FastAPI Version

0.89.1

### Python Version

3.11.0

**Comment by igomonteiro** [comment]
I'm having issues too, after updated the fastapi, the json encoder isn't working, i'm getting error by returning a list of a pydantic model or database model.

**Comment by igomonteiro** [comment, marked as answer]
Sorry, this is actually a FastAPI problem with SQLAlchemy, I don't know what changed in the last update of SQLAlchemy, but with version 1.4.46 of SQLAlchemy, json_encoder worked normally. But with the latest version of SQLAlchemy which was released yesterday it is not working.
